### ISSUE TITLE
LDAP External User Directory role mapping leads to query hangs and denial of service

### ISSUE THREAD

**Issue by mark-bb** [ClickHouse/ClickHouse#53708]
LDAP External User Directory with role mapping is configured against Microsoft Active Directory as described here:
https://clickhouse.com/docs/en/operations/external-authenticators/ldap
If some user is added to an AD group which doesn't have a corresponding role in ClickHouse after some previous query activity of this user, then subsequent TCP and HTTP queries like `SHOW ACCESS` / `SHOW CURRENT ROLES` / `SELECT user()` issued by this user start timeout (TCP) or even hang forever (HTTP ones). All subsequent queries from this user timeout or hang as well.
`netstat` shows a growing number of connections in a `CLOSE_WAIT` state afterwards which eventually leads to the whole ClickHouse system hang.

**It's reproduced on 22.12 at least and all 23 releases including the latest one on CentOS, RHEL, Ubuntu**

**How to reproduce**

* ClickHouse server version: 22.12 and all 23.x
* Interfaces: clickhouse-client, curl, JDBC driver
```
sudo cat /etc/clickhouse-server/config.d/ldap.xml
<clickhouse>

        <ldap_servers>
                <my_ad_server>
                        <host>somename.dcch.local</host>
                        <port>389</port>
                        <enable_tls>no</enable_tls>
                        <tls_require_cert>never</tls_require_cert>
                        <bind_dn>SOMENAME\{user_name}</bind_dn>
                        <user_dn_detection>
                                <base_dn>OU=CHILD,DC=somename,DC=dcch,DC=local</base_dn>
                                <search_filter>(&amp;(objectClass=user)(sAMAccountName={user_name}))</search_filter>
                        </user_dn_detection>
                </my_ad_server>
        </ldap_servers>

        <user_directories>
                <ldap>
                        <server>my_ad_server</server>
                        <role_mapping>
                                <base_dn>OU=CHILD,DC=somename,DC=dcch,DC=local</base_dn>
                                <attribute>CN</attribute>
                                <scope>subtree</scope>
                                <search_filter>(&amp;(objectClass=group)(member={user_dn}))</search_filter>
                        </role_mapping>
                </ldap>
        </user_directories>

</clickhouse>
```
1. Initially:
`ldapsearch` with the `(&(objectClass=user)(sAMAccountName=myuser))` filter returns:
```memberOf: CN=ch_role1,OU=CHILD,DC=somename,DC=dcch,DC=local```

The role `ch_role1` was created with the `CREATE ROLE ch_role1` statement and:
```
SELECT name, storage FROM system.roles

|name    |storage        |
|--------|---------------|
|ch_role1|local directory|
```
2. `myuser` successfully runs the following queries:
```
clickhouse-client -h localhost -u myuser --password xyz -q 'SELECT user() FORMAT PrettyCompact'
┌─currentUser()─┐
│ myuser        │
└───────────────┘

clickhouse-client -h localhost -u myuser --password xyz -q 'SHOW CURRENT ROLES FORMAT PrettyCompact'
┌─role_name─┬─with_admin_option─┬─is_default─┐
│ ch_role1  │                 0 │          1 │
└───────────┴───────────────────┴────────────┘
```

3. `myuser` is added to another AD group `ch_role2` which doesn't have a corresponding role in ClickHouse
`ldapsearch` returns:
```
memberOf: CN=ch_role2,OU=CHILD,DC=somename,DC=dcch,DC=local
memberOf: CN=ch_role1,OU=CHILD,DC=somename,DC=dcch,DC=local
```
4. The same query timed out
```
clickhouse-client -h localhost -u myuser --password xyz -q 'SELECT user()'

Code: 209. DB::NetException: Timeout exceeded while reading from socket (127.0.0.1:9000, 300000 ms). (SOCKET_TIMEOUT)
```
Whatever HTTP query from this user hangs forever afterwards:
```
echo 'SELECT 1' | curl "http://localhost:8123" --user "myuser:xyz" --data-binary @-
```
Subsequent queries from other users may or may not hang (it depends on version and release probably).

5. network connections in the `CLOSE_WAIT` state appear afterwards and their number grows until the whole ClickHouse server becomes unresponsive
```
netstat -an | grep CLOSE
tcp6       1      0 127.0.0.1:9000          127.0.0.1:59950         CLOSE_WAIT
tcp6       1      0 127.0.0.1:8123          127.0.0.1:38826         CLOSE_WAIT
```
6. The only way to make the ClickHouse server work normally again is to restart its service.

**Expected behavior**

User queries and ClickHouse server must not hang after changing user group membership in a LDAP server.

**Comment by Enmk** [comment]
Hi @mark-bb ! Do you have any LDAP-related exception in ClickHouse server logs?
Could you please share some excerpts, that would help a lot. LDAP is kind of not very trivial to test, so that might save some debugging time.

Thanks!

**Comment by mark-bb** [comment]
There is nothing in the clickhouse logs even on the trace level on this. It seems that setting of clickhouse logging levels doesn't affect logging of ldap communications at all.

**Comment by alezzqz** [comment]
> Hi @mark-bb ! Do you have any LDAP-related exception in ClickHouse server logs? Could you please share some excerpts, that would help a lot. LDAP is kind of not very trivial to test, so that might save some debugging time.

Hi @Enmk 
It's pretty easy to reproduce the situation using description above. Actually there's deadlock occurs in LDAPAccessStorage on updating roles. Seems like a bug. I've got a patch fixing the problem.

**Comment by mark-bb** [comment]
Hi All,

Great news about fixing the problem!
Is there any information on when the corresponding fix will be available for download on 23.3 and other lts builds (if any)?
Thanks!

**Comment by Trifonov-Artem** [comment]
We have this error on LTS versions 23.8.9.54 and 24.3.1.2672
